src/combinators/mod.rs

```diff
@@ -1,164 +1,123 @@
-///     assert_eq!(buf.len(), 96);
-///     assert_eq!(len, 4);
-/// }
-///
-/// assert_eq!(&buf[..4], &b"abcd"[..]);
-/// ```
-///
-/// For byte slices `std::io::Cursor` provides more features and allows to retrieve the original
-/// slice
-///
-/// ```rust
-/// #[cfg(feature = "std")]
-/// use std::io::Cursor;
-/// #[cfg(not(feature = "std"))]
-/// use cookie_factory::lib::std::io::{Cursor, Write};
-///
-/// use cookie_factory::string;
-///
-/// let mut buf = [0u8; 100];
-///
-/// let mut cursor = Cursor::new(&mut buf[..]);
-/// cursor = string("abcd")(cursor).unwrap();
-/// assert_eq!(cursor.position(), 4);
-/// let buf = cursor.into_inner();
-///
-/// assert_eq!(&buf[..4], &b"abcd"[..]);
-/// ```
-pub struct WriteCounter<W>(W, u64);
+    pub fn set_position(&mut self, position: u64) {
+        self.position = position;
+    }
 
-impl<W: Write> WriteCounter<W> {
-    /// Create a new `WriteCounter` around `w`
-    pub fn new(w: W) -> Self {
-        WriteCounter(w, 0)
+    pub fn get_ref(&self) -> &W {
+        &self.write
     }
 
-    /// Returns the amount of bytes written so far
-    pub fn position(&self) -> u64 {
-        self.1
+    pub fn get_mut(&mut self) -> &mut W {
+        &mut self.write
     }
 
-    /// Consumes the `WriteCounter` and returns the contained `Write` and the amount of bytes
-    /// written
     pub fn into_inner(self) -> (W, u64) {
-        (self.0, self.1)
+        (self.write, self.position)
     }
 }
 
-impl<W: Write> Write for WriteCounter<W> {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        let written = self.0.write(buf)?;
-        self.1 += written as u64;
-        Ok(written)
+impl<W: Write> Write for WriteContext<W> {
+    fn write(&mut self, data: &[u8]) -> crate::lib::std::io::Result<usize> {
+        let amt = self.write.write(data)?;
+        self.position += amt as u64;
+        Ok(amt)
     }
 
-    // Our minimal Write trait has no flush()
     #[cfg(feature = "std")]
     fn flush(&mut self) -> io::Result<()> {
-        self.0.flush()
+        self.write.flush()
     }
 }
 
-impl<W: Skip> Skip for WriteCounter<W> {
-    fn skip(mut self, len: usize) -> Result<Self, GenError> {
-        self.0 = self.0.skip(len)?;
-        self.1 += len as u64;
-        Ok(self)
-    }
-}
-
-impl<W: Seek> io::Seek for WriteCounter<W> {
+impl<W: Seek> io::Seek for WriteContext<W> {
     fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
-        let old_pos = self.0.seek(SeekFrom::Current(0))?;
-        let new_pos = self.0.seek(pos)?;
+        let old_pos = self.write.seek(SeekFrom::Current(0))?;
+        let new_pos = self.write.seek(pos)?;
         if new_pos >= old_pos {
-            self.1 += new_pos - old_pos;
+            self.position += new_pos - old_pos;
         } else {
-            self.1 -= old_pos - new_pos;
+            self.position -= old_pos - new_pos;
         }
         Ok(new_pos)
     }
 }
 
+pub fn gen<W: Write, F: SerializeFn<W>>(f: F, w: W) -> GenResult<W> {
```

Comment:
I was reluctant at first wrt this and gen_simple, but they can help keep things concise so fine by me

Reply:
It's the prize we have to pay for having the `WriteContext` unfortunately ;)